hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java

```diff
@@ -192,125 +166,173 @@ public void init(NodeEngine nodeEngine, Properties properties) {
+        } else {
+            membershipManagerCompat.removeMember(suspectedAddress, suspectedUuid, reason);
         }
-        MemberImpl member = getMember(target);
-        String memberUuid = member != null ? member.getUuid() : null;
-        Collection<MemberImpl> members = getMemberImpls();
-        MemberInfoUpdateOperation op = new MemberInfoUpdateOperation(memberUuid, createMemberInfoList(members),
-                                                                     clusterClock.getClusterTime(), null, false);
-        nodeEngine.getOperationService().send(op, target);
     }
 
-    public void removeAddress(Address deadAddress, String uuid, String reason) {
+    public void handleMasterConfirmation(MembersViewMetadata membersViewMetadata, long timestamp) {
         lock.lock();
         try {
-            MemberImpl member = getMember(deadAddress);
-            if (member == null || !uuid.equals(member.getUuid())) {
-                if (logger.isFineEnabled()) {
-                    logger.fine("Cannot remove " + deadAddress + ", either member is not present "
-                            + "or uuid is not matching. Uuid: " + uuid + ", member: " + member);
-                }
+            if (!isJoined()) {
+                logger.warning("Ignoring heartbeat of sender: " + membersViewMetadata + " because not joined!");
                 return;
             }
-            doRemoveAddress(deadAddress, reason, true);
+
+            Address endpoint = membersViewMetadata.getMemberAddress();
+            MemberImpl member = membershipManager.getMember(endpoint, membersViewMetadata.getMemberUuid());
+            if (member == null) {
+                if (getClusterVersion().isGreaterOrEqual(Versions.V3_9)) {
+                    if (!isMaster()) {
+                        logger.warning(endpoint + " has sent MasterConfirmation with " + membersViewMetadata
+                                + ", but this node is not master!");
+                        return;
+                    }
+
+                    if (clusterJoinManager.isMastershipClaimInProgress()) {
+                        // this can be a new member I have discovered...
+                        return;
+                    }
+
+                    logger.warning(endpoint + " has sent MasterConfirmation with " + membersViewMetadata
+                            + ", but it is not a member of this cluster!");
+                    // This guy knows me as its master but I am not. I should explicitly tell it to remove me from its cluster.
+                    // It should suspect me so that it can move on.
+                    // IMPORTANT: I should not tell it to remove me from cluster while I am trying to claim my mastership.
+
+                    sendExplicitSuspicion(membersViewMetadata);
+
+                    for (Member m : getMembers(NON_LOCAL_MEMBER_SELECTOR)) {
+                        sendExplicitSuspicionTrigger(m.getAddress(), membersViewMetadata);
```

Comment:
Just checking - this means that all members in the cluster (that master knows of) will send an explicit suspicion to the faulty node (the one that sent the master confirmation to this node)? So that the faulty node will suspect all members of this cluster?